### ISSUE TITLE
NER using huggingface VS spacy transformers (inference only)

### ISSUE THREAD

**Issue by jankrepl** [BlueBrain/Search#307]
## 🚀 Feature
Currently, we are using `spacy` NER models. We might want to try using a pretrained NER model from `transformers`. Note that this ticket only talks about inference and possibly integrating it in our code base.

Note that `space==3.*` is actually tightly integrated with transformers so we should probably look into that too.

## Motivation
* Lower lever understanding of what the model is doing - possibly we could extract confidence scores without any hacks
* Possibly better performance
* More customizable

## Alternatives
Just use the `spacy` models as we do now.

## Additional context
* When it comes to pretraininig/finetuning a custom transformer NER I would suggest creating a totally separate issue
* The same applies to evaluation
* It might help us tackle #295 by understanding what other people/projects are doing

**Comment by jankrepl** [comment]
I did some research. Let me summarize:

###  NER in transformers
* **Extremely simple** - it is a special case of the **token classification** task
* **How does one go from token embeddings to NER?** The prototype can bee seen here [transformers.modeling_bert.BertForTokenClassification](https://github.com/huggingface/transformers/blob/0486ccdd3db82ef4b1f636f638f5744a4dcc2d44/src/transformers/models/bert/modeling_bert.py#L1638). One literally just adds a dropout and linear layer as the head and then applies it to each token. The `out_features` of the `torch.nn.Linear` is going to be equal to the number of entity types.
* **Can I get scores?** Yes, each token has logits (probabilities after softmax) over all entitiy types
* **Are there pretrained models already? Y**es - [here](https://huggingface.co/models?filter=token-classification).
* **Are there any problems?** If built upon BERT then there is going to be wordpiece-tokenization! 

### Spacy 3 + Transformers + NER
* Everything is implemented inside of a separate repo https://github.com/explosion/spacy-transformers/
* It seems like Spacy takes the word-piece tokens (coming from the `transformers`) and then "aligns" them so to the tokens of it's own tokenizer. The spac tokens embeddings are then just sums of the wordpiece token embeddings.[ Read more](https://spacy.io/api/transformer).
* They use the `thinc` library to do deep learning related things

# Actions and to discuss
* IMO we could use raw `transformers`, however, that would require our labeled data to be compatible with the BERT's wordpiece tokenizer. See [prodigy docs on this](https://prodi.gy/docs/named-entity-recognition#transformers-tokenizers).
*  If I am not really mistaken we do not necessarily care about the tokenization itself, right? Note that even if an entity is split up into multiple subwords the neural network should just predict `B-ETYPE`, `I-ETYPE`, ..., `I-ETYPE` and we can just group them together. See `grouped_entities` [here](https://github.com/huggingface/transformers/blob/0486ccdd3db82ef4b1f636f638f5744a4dcc2d44/src/transformers/pipelines/token_classification.py#L87).
* **Can we have both spacy and raw transformers?** IMO not unless we recreate some alignment logic as `spacy-transformers`.

**Comment by FrancescoCasalegno** [comment]
Some comments on the "Pros and cons" table above.
- **Pretraining**
   - `spaCy` does support pre-training natively, there's even a [`spacy pretrain`](https://spacy.io/usage/v2-1#pretraining).
   - As an alternative, one could still perform the pre-training outside of `spaCy`, checkpoint the resulting model, and then load it in `spaCy` (which is pretty much the experiment implemented in #328)
- **Possibility to look under the hood**
   - I can confirm that also in my experience `spaCy` has so many layers of abstraction and indirection (see `thinc` and its docs which are really tricky to navigate), that it becomes hard to understand what is _really_ happening under the hood. :(
   - Even though this is not necessarily nice, we should also try to understand how often and in which case should we feel the need of looking under the hood? I.e. is this a must-have or just a nice-to-have feature?
   - 
- **Can achieve state of the art?**
  -  Let's ask `spaCy` developers (i) why/whether are they using this [transition-based model](https://arxiv.org/pdf/1603.01360.pdf) rather than any other model that claims SotA for ner tasks; (ii) whether it is possible to load other SotA architectures implemented e.g. in `torch` using `spaCy`.
- **Adding manual rules**
   - Using `huggingface` we could still run first a NER using `huggingface`  and then the EntityRuler written in `spaCy`.
- **Confidence scores**
   - Aggregation seems tricky in `torch`, and probably impossible atm with `spaCy`.
   - Seems like a nice-to-have rather than a must-have requirement.
- **Scalability**
   - `spaCy` insists on being blazing fast and scalable — but what if we have different `tok2vec` (or `transformer`) base models for different pipeline stages?
   - If we use caching, it seems like a nice-to-have rather than a must-have requirement.
- ** **

**Comment by FrancescoCasalegno** [comment]
## 2021-04-07 Decision

For the time being, the incremental solution that implies the least amount of re-work is to go for `spaCy`.
The following list collects all known risks/shortcomings of this solution and tries to identify ways to address them.

## Known Unknowns
1. **Pre-training**. 
Currently we use the scripts of [`data_and_models/pipelines/sentence_embedding/scripts/`](https://github.com/BlueBrain/Search/tree/master/data_and_models/pipelines/sentence_embedding/scripts) to pre-train the weights of the transformer model that is then used as a backbone and fine-tuned to generate our NER model. Those scripts are mainly  (a) a copy-paste of the `huggingface` script to run the LM (language modelling) pre-training ([here](https://github.com/huggingface/transformers/blob/v3.4.0/examples/language-modeling/run_language_modeling.py)) and (b) of  the `sentence-transformers` script to run the NLI+STS fine-tuning ([here](https://github.com/UKPLab/sentence-transformers/blob/v0.3.9/examples/training/other/training_multi-task.py)). 
But there's also a [`spacy pretrain`](https://spacy.io/api/cli#pretrain) that lets you pre-train a model using a self-superivised LM task ("LMAO"–Language Modelling with Approximate Outputs). The docs say that this allows to pre-train a `Tok2Vec` component, but it's not clear if they also support doing it for the alternative `Transformer` component.
     - [ ] Investigate if `spacy pretrain` also supports pre-training a `Transformer` component (open issue on their GH if needed). If Yes, determine if we can use this command instead of the aforementioned script (a) from `huggingface`.
2. **NER head model architecture**.
While both `huggingface` and `spacy` approaches to NER (see #318 and #319, respectively) use a transformer backbone model, the two approaches differ in the way the structure the head model. In fact, `huggingface` uses `torch` and therefore supports whatever you want to implement, while `spacy` seems to only have the default [`TransitionBasedParser`](https://spacy.io/api/architectures#TransitionBasedParser) architecture.
    - [ ] Present in a tech talk how this architecture works. Is it still what described in [Lample 2016](https://arxiv.org/pdf/1603.01360.pdf)? If Yes, why do they stick to such an old paper while there are now lots of various approaches in sota papers? Ask `spacy` devs if needed.
    - [ ] Investigate how/whether one can register and load a head model written in `torch` into `spacy`. Are there examples of that? Ask `spacy` devs if needed.
3. **Reproducibility**
For us one of the strong selling points of `spacy` in the past was that it allowed reproducible training results. It is important to know if that's still the case so that we know if `dvc` can be safely used to track the pipeline stages.
    - [ ] Are `spacy` training results still bitwise reproducible even if we use transformers? And also if we train on GPU?
 4. **`config.cfg`**
   In `spacy 3.0` they introduced this `config.cfg` file that makes all training hyper-parmaters and settings explicit. But even if (i) there are various automated ways to generate this config file and (ii) the Docs do have a section explaining how this file works ([here](https://spacy.io/usage/training#config)), experience clearly showed us that it's really hard to understand all details of it. At the same time, a deep understanding of this file is needed for us to be able to effectively use `spacy` on our project.
     - [ ] Compare our training script with the one that `prodigy` generates. --> 📌 #309 
     - [ ]  Present in a tech talk all the details and info available on the `config.cfg` file. This should also include advanced features such as an example of registered function etc.

**Comment by pafonta** [comment]
Inside point 2:

> while spacy seems to only have the default TransitionBasedParser architecture

According to the documentation of `spaCy 3`, that's not the case.

Indeed, one seems to be able to use any architecture, including any `nn.Module` from `PyTorch`:

https://spacy.io/usage/layers-architectures#frameworks

**Comment by Stannislav** [comment]
I think it's a good summary.

Off the top of my head, some might be off-topic:
- Would the performance of a huggingface-based model be comparable to previous results for all entity types?
- Training huggingface needs to be explored more (which base model? which hyperparameters for training?)
- Both huggingface and spacy:
  - How to add a new entity type?
  - One model per entity type or not?
  - How to get new annotations for new entity types?

**Comment by jankrepl** [comment]
One other thing to consider, our code is currently written to work with`spacy` NER. So if we decide to fully switch to `transformers` we would need to spend some time adjusting `bluesearch`.

**Comment by EmilieDel** [comment]
I agree with all the points from everyone!

**Comment by FrancescoCasalegno** [comment]
All To Do's listed in [this comment](https://github.com/BlueBrain/Search/issues/307#issuecomment-814778506) have now corresponding issues to be addressed.

Closing.
